File: docs/vsto/codesnippet/CSharp/Trin_VstcoreActionsPaneWordCS/ActionsControl.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Text;
using System.Windows.Forms;

namespace Trin_VstcoreActionsPaneWordCS
{
    partial class ActionsControl : UserControl
    {
        //---------------------------------------------------------------------
        //<Snippet1>
        private void ActionsControl_Load(object sender, EventArgs e)
        {
            this.suppliersTableAdapter.Fill(this.northwindDataSet.Suppliers);
            this.productsTableAdapter.Fill(this.northwindDataSet.Products);
        }
        //</Snippet1>


        //---------------------------------------------------------------------
        //<Snippet2>
        static void SetHeadings(Microsoft.Office.Interop.Word.Cell tblCell, string text)
        {
            tblCell.Range.Text = text;
            tblCell.Range.Font.Bold = 1;
            tblCell.Range.ParagraphFormat.Alignment = 
                Microsoft.Office.Interop.Word.WdParagraphAlignment.wdAlignParagraphCenter;
        }
        //</Snippet2>


        //---------------------------------------------------------------------
        //<Snippet3>
        private void AddData(System.Data.DataRow row, string companyName)
        {
            object missing = System.Type.Missing;

            // Create a table if it doesn't already exist.
            if (Globals.ThisDocument.Tables.Count == 0)
            {
                try
                {
                    // Create a table.
                    Microsoft.Office.Interop.Word.Table tbl = Globals.ThisDocument.Tables.Add
                        (Globals.ThisDocument.Application.Selection.Range, 1, 4, ref missing, ref missing);

                    // Insert headings.
                    SetHeadings(tbl.Cell(1, 1), "Company Name");
                    SetHeadings(tbl.Cell(1, 2), "Product Name");
                    SetHeadings(tbl.Cell(1, 3), "Quantity");
                    SetHeadings(tbl.Cell(1, 4), "Unit Price");
                }
                catch (Exception ex)
                {
                    MessageBox.Show("Problem creating Products table: " + ex.Message, 
                        "Actions Pane", MessageBoxButtons.OK, MessageBoxIcon.Error);
                }
            }

            // Add data from data row to the table.
            Microsoft.Office.Interop.Word.Selection selection = Globals.ThisDocument.Application.Selection;

            if (selection.Tables.Count > 0)
            {
                Microsoft.Office.Interop.Word.Row newRow = Globals.ThisDocument.Tables[1].Rows.Add(ref missing);

                newRow.Range.Font.Bold = 0;

                newRow.Range.ParagraphFormat.Alignment = 
                    Microsoft.Office.Interop.Word.WdParagraphAlignment.wdAlignParagraphLeft;

                newRow.Cells[4].Range.ParagraphFormat.Alignment =
                    Microsoft.Office.Interop.Word.WdParagraphAlignment.wdAlignParagraphRight;

                newRow.Cells[1].Range.Text = companyName;
                newRow.Cells[2].Range.Text = row["ProductName"].ToString();
                newRow.Cells[3].Range.Text = row["QuantityPerUnit"].ToString();
                newRow.Cells[4].Range.Text = Math.Round(Convert.ToDouble(row["UnitPrice"])).ToString("#,##0.00");
            }
            else
            {
                MessageBox.Show("Cursor must be within a table.", 
                    "Actions Pane", MessageBoxButtons.OK, MessageBoxIcon.Error);
            }
        }
        //</Snippet3>
        

        //---------------------------------------------------------------------
        //<Snippet4>
        private void Insert_Click(object sender, System.EventArgs e)
        {
            System.Data.DataTable tbl = northwindDataSet.Products;
            System.Data.DataRow[] rows;

            // Check if a product is selected.
            if (this.productNameListBox.SelectedIndex >= 0)
            {
                System.Data.DataRowView productRow = (System.Data.DataRowView)this.productNameListBox.SelectedItem;

                string product = productRow.Row["ProductName"].ToString();
                string company = this.companyNameComboBox.Text;

                // Return the data row from the selected product.
                rows = tbl.Select("[ProductName] = '" + product.Replace("'", "''") + "'");

                this.AddData(rows[0], company);
            }
            else
            {
                MessageBox.Show("Please select a product.", "Actions Pane", MessageBoxButtons.OK);
            }
        }
        //</Snippet4>


        //---------------------------------------------------------------------
        public ActionsControl()
        {
            InitializeComponent();

            //<Snippet5>
            this.Insert.Click += new EventHandler(Insert_Click);
            //</Snippet5>

            //<Snippet33>
            this.Load += new EventHandler(ActionsControl_Load);
            //</Snippet33>
        }
    }
}

File: docs/vsto/codesnippet/VisualBasic/Trin_VstcoreActionsPaneWordVB/ActionsControl.vb
Public Class ActionsControl


    '--------------------------------------------------------------------------
    '<Snippet1>
    Private Sub ActionsControl_Load(ByVal sender As Object, ByVal e As System.EventArgs) _
        Handles Me.Load

        Me.SuppliersTableAdapter.Fill(Me.NorthwindDataSet.Suppliers)
        Me.ProductsTableAdapter.Fill(Me.NorthwindDataSet.Products)
    End Sub
    '</Snippet1>


    '--------------------------------------------------------------------------
    '<Snippet2>
    Shared Sub SetHeadings(ByVal tblCell As Word.Cell, ByVal text As String)

        With tblCell.Range
            .Text = text
            .Font.Bold = True
            .ParagraphFormat.Alignment = Word.WdParagraphAlignment.wdAlignParagraphCenter
        End With
    End Sub
    '</Snippet2>


    '--------------------------------------------------------------------------
    '<Snippet3>
    Private Sub AddData(ByVal row As System.Data.DataRow, ByVal companyName As String)

        ' Create a table if it doesn't already exist.
        If Globals.ThisDocument.Tables.Count = 0 Then

            Try
                ' Create a table. 
                Dim tbl As Word.Table = Globals.ThisDocument.Tables.Add( _
                    Globals.ThisDocument.Application.Selection.Range, 1, 4)

                ' Insert headings.
                SetHeadings(tbl.Cell(1, 1), "Company Name")
                SetHeadings(tbl.Cell(1, 2), "Product Name")
                SetHeadings(tbl.Cell(1, 3), "Quantity")
                SetHeadings(tbl.Cell(1, 4), "Unit Price")

            Catch ex As Exception
                MessageBox.Show("Problem creating Products table: " & ex.Message, _
                    "Actions Pane", MessageBoxButtons.OK, MessageBoxIcon.Error)
            End Try
        End If

        ' Add data from data row to the table.
        Dim selection As Word.Selection = Globals.ThisDocument.Application.Selection

        If selection.Tables.Count > 0 Then

            Dim newRow As Word.Row = Globals.ThisDocument.Tables(1).Rows.Add()
            With newRow
                .Range.Font.Bold = False
                .Range.ParagraphFormat.Alignment = Word.WdParagraphAlignment.wdAlignParagraphLeft
                .Cells(4).Range.ParagraphFormat.Alignment = Word.WdParagraphAlignment.wdAlignParagraphRight
                .Cells(1).Range.Text = companyName
                .Cells(2).Range.Text = row.Item("ProductName").ToString
                .Cells(3).Range.Text = row.Item("QuantityPerUnit").ToString
                .Cells(4).Range.Text = Math.Round(row.Item("UnitPrice"), 2)
            End With

        Else
            MessageBox.Show("Cursor must be within a table.", _
                "Actions Pane", MessageBoxButtons.OK, MessageBoxIcon.Error)
        End If
    End Sub
    '</Snippet3>


    '--------------------------------------------------------------------------
    '<Snippet4>
    Private Sub Insert_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Insert.Click

        Dim tbl As System.Data.DataTable = NorthwindDataSet.Products
        Dim rows() As System.Data.DataRow

        ' Check if a product is selected.
        If Not Me.ProductNameListBox.SelectedIndex < 0 Then

            Dim product As String = Me.ProductNameListBox.SelectedItem.Row.Item("ProductName")
            Dim company As String = Me.CompanyNameComboBox.Text

            ' Return the data row from the selected Product in the list box.
            rows = tbl.Select("[ProductName] = '" & product.Replace("'", "''") & "'")

            AddData(rows(0), company)

        Else
            MessageBox.Show("Please select a product.", "Actions Pane", MessageBoxButtons.OK)
        End If
    End Sub
    '</Snippet4>

End Class
